03-Custom-Actions-In-Python/02-An-Initial-Custom-Action.ipynb

```diff
@@ -0,0 +1,247 @@
+{
```

Comment:
<p>How does the custom action correctly grab the timestep? You mention </p><p>"(If you are wondering how to access simulation data, there is a mechanism for that which we will go over further down). "</p><p>How would a user know which keywords will work and which wont?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Reply:
<p>+1 to this confusion</p>

Reply:
<p>Yeah, I'm still confused on this. Is there a way to list all the variables that can be used in a custom action? Showing the user that could be really helpful.</p>

Reply:
@cbkerr @jennyfothergill I am somewhat confused by what is meant here. The timestep is passed in as a function argument so the action isn't responsible for querying that. Likewise, the only attribute we add is `_state`. I mention that operations expose their actions attributes which means trigger should not be an attribute name, but their isn't really a list just `_state` and whatever else is added in a custom defined `attach` method.

Reply:
<p>I guess what I'm saying is if the timestep variable in the PrintTimestep class was changed to something nonsensical, would this action still work or does the name matter? Does it work because there is a named property of the Simulation class called timestep? How could a user know what named properties they can use in their custom action? Would <code>dir(sim)</code>  or <code>dir(sim.state)</code>  give them a place to start? Is there a place in the docs we could link?</p>

Reply:
@jennyfothergill @cbkerr The first argument to `act` is the timestep. If you defined your act like this:
```def act(self, data):```
, then the caller will still pass the timestep in the first argument and `data` will be the timestep.

Reply:
<p>OK, so the variable name in the action doesn't matter, is that right? I still don't understand how it correctly grabs the timestep. It looks to me that the timestep is never passed to the action-- the action is added as a custom writer and then somehow grabs the timestep. How would a user go about printing a value other than timestep?</p><p> </p>

Reply:
@jennyfothergill The C++ code that calls the custom action's act method passes the timestep as a positional argument:
https://github.com/glotzerlab/hoomd-blue/blob/898a8e4b179ffb9a51aa97f0a9ce6b9453191753/hoomd/PythonAnalyzer.cc#L12-L15

This is just the way HOOMD is designed. When you write code that extends another code, you need to follow the same API. There is nothing special about HOOMD in this regard. The API defines the method names that programmers and the rest of the code base expect to exist and the arguments that they take.

Reply:
It might be enough to link to the API docs here. Once I found [this](https://hoomd-blue.readthedocs.io/en/latest/module-hoomd-custom.html#hoomd.custom.Action.act) it helped.